From 0ca24bbc22324e90c873d1196d274d21a50da42f Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Mon, 8 May 2023 17:48:07 +0200
Subject: [PATCH 1002/1006] Load user data handler from function so we can
 access it early

We need to acces user-data from "main" phase, so we need to have a way
to load user-data handler after user-data is saved.
---
 lib/tiny-cloud/cloud/alpine/init | 18 ++++++++++++++++++
 lib/tiny-cloud/init              | 12 ++++++++----
 2 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/lib/tiny-cloud/cloud/alpine/init b/lib/tiny-cloud/cloud/alpine/init
index 2435c75..b146b6c 100644
--- a/lib/tiny-cloud/cloud/alpine/init
+++ b/lib/tiny-cloud/cloud/alpine/init
@@ -14,7 +14,20 @@ replace_word() {
     done
 }
 
+# usage: insert_after <where> <what> <list>...
+insert_after() {
+    local search="$1" addition="$2"
+    shift 2
+    for i in "$@"; do
+        echo "$i"
+        if [ "$i" = "$search" ]; then
+            echo "$addition"
+        fi
+    done
+}
+
 INIT_ACTIONS_EARLY="$(replace_word set_default_interfaces set_network_interfaces $INIT_ACTIONS_EARLY)"
+INIT_ACTIONS_MAIN="$(insert_after save_userdata load_userdata_handler $INIT_ACTIONS_MAIN)"
 
 add_once() {
     local file="$1"
@@ -48,3 +61,8 @@ init__set_network_interfaces() {
         set_resolv_conf
     fi
 }
+
+init__load_userdata_handler() {
+    load_userdata_handler
+}
+
diff --git a/lib/tiny-cloud/init b/lib/tiny-cloud/init
index 580ab2d..a449c14 100644
--- a/lib/tiny-cloud/init
+++ b/lib/tiny-cloud/init
@@ -276,8 +276,12 @@ userdata_type() {
     esac
 }
 
-USERDATA_TYPE="$(userdata_type)"
-if [ -f "$LIBDIR/tiny-cloud/user-data/$USERDATA_TYPE" ]; then
-    . "$LIBDIR/tiny-cloud/user-data/$USERDATA_TYPE"
-fi
+load_userdata_handler() {
+    USERDATA_TYPE="$(userdata_type)"
+    if [ -f "$LIBDIR/tiny-cloud/user-data/$USERDATA_TYPE" ]; then
+        . "$LIBDIR/tiny-cloud/user-data/$USERDATA_TYPE"
+    fi
+}
+
+load_userdata_handler
 # TODO: some indication that the user-data type is unsupported?
-- 
2.40.1

